cirq/protocols/approx_equality.py

```diff
@@ -0,0 +1,172 @@
+            details.
+        abs_tol: The minimum absolute tolerance. See math.isclose()
+            documentation for details.
+
+        Returns:
+            True if objects are approximately equal, False otherwise. Returns
+            NotImplemented when approximate equality is not implemented for
+            given types.
+        """
+        pass
+
+
+def approx_eq(
+        val: Any,
+        other: Any,
+        rel_tol: float = 1e-09,
+        abs_tol: float = 0.0) -> bool:
+    """Approximately compares two objects.
+
+    If `val` implements SupportsApproxEquality protocol then it is invoked and
+    takes precedence over all other checks:
+     - For primitive numeric types `int` and `float` approximate equality is
+       delegated to math.isclose().
+     - For complex primitive type the real and imaginary parts are treated
+       independently and compared using math.isclose().
+     - For `val` and `other` both iterable, consecutive elements are compared
+       recursively. Types of `val` and `other` does not necessarily needs to
+       match each other. They just need to be iterable and have the same
+       structure.
+
+    Args:
+        val: Source object for approximate comparison.
+        other: Target object for approximate comparison.
+        rel_tol: The relative tolerance. See math.isclose() documentation for
+            details.
+        abs_tol: The minimum absolute tolerance. See math.isclose()
+            documentation for details.
+
+    Returns:
+        True if objects are approximately equal, False otherwise. Returns
+        NotImplemented when approximate equality is not implemented for given
+        types.
+    """
+
+    # Check if object defines approximate equality via _approx_eq_. This takes
+    # precedence over all other overloads.
+    approx_eq_getter = getattr(val, '_approx_eq_', None)
+    if approx_eq_getter is not None:
+        return approx_eq_getter(other, rel_tol, abs_tol)
+
+    val_t = type(val)
+    other_t = type(other)
+
+    # Fallback to math.isclose() for primitive types, which is symmetric
+    # contrary to numpy.isclose().
+    if val_t in [int, float]:
+        if other_t not in [int, float]:
+            return NotImplemented
+        return math.isclose(val, other, rel_tol=rel_tol, abs_tol=abs_tol)
+
+    # For complex types, treat real and imaginary parts independently.
+    if val_t == complex:
+        if val_t != other_t:
+            return NotImplemented
+        return math.isclose(
+            val.real,
+            other.real,
+            rel_tol=rel_tol,
+            abs_tol=abs_tol
+        ) and math.isclose(
+            val.imag,
+            other.imag,
+            rel_tol=rel_tol,
+            abs_tol=abs_tol
+        )
+
+    # Try to compare source and target recursively, assuming they're iterable.
+    return _approx_eq_iterables(val, other, rel_tol=rel_tol, abs_tol=abs_tol)
```

Comment:
Another fallback we need, after everything else fails including the iterability check, is the fallback to `__eq__`.

Reply:
Done.